### ISSUE TITLE
Implement middleware for checking if Identity has been created

### ISSUE THREAD

**Issue by zupzup** [BitcreditProtocol/E-Bill#193]
If no identity has been created yet, all endpoints except the identity-related ones should return `Not acceptable`.

Currently this is implemented by checking for the `identity` file to exist and is implemented in all handlers manually.

Create an `is_identity_initialized` middleware, or something, which does a proper check and include it everywhere necessary.

Use a rocket Fairing (https://rocket.rs/guide/v0.5/fairings/) to implement this, so the check doesn't have to be in every handler.

**Comment by zupzup** [comment]
@mubarak23 
Concretely, this is about these calls in `web/handler`

```rust
    if !state.identity_service.identity_exists().await {
        return Err(service::Error::PreconditionFailed);
    }
```

They check if the identity has been created and if not, return an error. This prevents the application to respond if it wasn't initialized.
The goal is to add a Rocket Fairing, which encapsulates this in a middleware, so we don't have that call in every single handler.

**Comment by mubarak23** [comment]
> @mubarak23 Concretely, this is about these calls in `web/handler`
> 
> ```rust
>     if !state.identity_service.identity_exists().await {
>         return Err(service::Error::PreconditionFailed);
>     }
> ```
> 
> They check if the identity has been created and if not, return an error. This prevents the application to respond if it wasn't initialized. The goal is to add a Rocket Fairing, which encapsulates this in a middleware, so we don't have that call in every single handler.

Alright, 

on `on_response` callback we are checking if `identity_exists()` i.e 

  ```
  async fn on_response<'r>(&self, request: &'r Request<'_>, response: &mut Response<'r>) {
        if let Some(context) = request.rocket().state::<ServiceContext>() {
            if !context.identity_service.identity_exists().await {
                return response.set_status(Status::NotAcceptable);
            }
        }
    }
```

**Comment by zupzup** [comment]
> > @mubarak23 Concretely, this is about these calls in `web/handler`
> > ```rust
> >     if !state.identity_service.identity_exists().await {
> >         return Err(service::Error::PreconditionFailed);
> >     }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > They check if the identity has been created and if not, return an error. This prevents the application to respond if it wasn't initialized. The goal is to add a Rocket Fairing, which encapsulates this in a middleware, so we don't have that call in every single handler.
> 
> Alright,
> 
> on `on_response` callback we are checking if `identity_exists()` i.e
> 
> ```
> async fn on_response<'r>(&self, request: &'r Request<'_>, response: &mut Response<'r>) {
>       if let Some(context) = request.rocket().state::<ServiceContext>() {
>           if !context.identity_service.identity_exists().await {
>               return response.set_status(Status::NotAcceptable);
>           }
>       }
>   }
> ```

No, it would have to be `on_request`, since we want to check it immediately when the request is received and then return an error, if the identity doesn't exist.

`on_response` is called at the very end of a request.

**Comment by mubarak23** [comment]
> > > @mubarak23 Concretely, this is about these calls in `web/handler`
> > > ```rust
> > >     if !state.identity_service.identity_exists().await {
> > >         return Err(service::Error::PreconditionFailed);
> > >     }
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > They check if the identity has been created and if not, return an error. This prevents the application to respond if it wasn't initialized. The goal is to add a Rocket Fairing, which encapsulates this in a middleware, so we don't have that call in every single handler.
> > 
> > 
> > Alright,
> > on `on_response` callback we are checking if `identity_exists()` i.e
> > ```
> > async fn on_response<'r>(&self, request: &'r Request<'_>, response: &mut Response<'r>) {
> >       if let Some(context) = request.rocket().state::<ServiceContext>() {
> >           if !context.identity_service.identity_exists().await {
> >               return response.set_status(Status::NotAcceptable);
> >           }
> >       }
> >   }
> > ```
> 
> No, it would have to be `on_request`, since we want to check it immediately when the request is received and then return an error, if the identity doesn't exist.
> 
> `on_response` is called at the very end of a request.
According to Fairing docs, 
on_request does not allow us to handle respond directly to the request or about. but it suggest we used `request guards`

**Comment by zupzup** [comment]
> > > > @mubarak23 Concretely, this is about these calls in `web/handler`
> > > > ```rust
> > > >     if !state.identity_service.identity_exists().await {
> > > >         return Err(service::Error::PreconditionFailed);
> > > >     }
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > They check if the identity has been created and if not, return an error. This prevents the application to respond if it wasn't initialized. The goal is to add a Rocket Fairing, which encapsulates this in a middleware, so we don't have that call in every single handler.
> > > 
> > > 
> > > Alright,
> > > on `on_response` callback we are checking if `identity_exists()` i.e
> > > ```
> > > async fn on_response<'r>(&self, request: &'r Request<'_>, response: &mut Response<'r>) {
> > >       if let Some(context) = request.rocket().state::<ServiceContext>() {
> > >           if !context.identity_service.identity_exists().await {
> > >               return response.set_status(Status::NotAcceptable);
> > >           }
> > >       }
> > >   }
> > > ```
> > 
> > 
> > No, it would have to be `on_request`, since we want to check it immediately when the request is received and then return an error, if the identity doesn't exist.
> > `on_response` is called at the very end of a request.
> > According to Fairing docs,
> > on_request does not allow us to handle respond directly to the request or about. but it suggest we used `request guards`

Ah, interesting.
Yes, then I guess a `request guard` would be the tool of choice. :+1:

**Comment by mtbitcr** [comment]
Hey @mubarak23 . Any progress here?

**Comment by mubarak23** [comment]
> Hey @mubarak23 . Any progress here?

yaah, am having some errors using the request guard, i will share the errors

**Comment by mubarak23** [comment]
@zupzup i have a PR for this issue #333 for review

**Comment by zupzup** [comment]
@mtbitcr this is merged ✅

**Comment by mubarak23** [comment]
@zupzup base on this merged PR, what issue do you think is best for me to pick up next.

**Comment by mtbitcr** [comment]
@mubarak23 Thank you for your contribution! You will be rewarded with 750 tokens at the beginning of the next quarter.
